tests/src/Interop/COM/NativeServer/StringTesting.h

```diff
@@ -0,0 +1,327 @@
+
+class DECLSPEC_UUID("C73C83E8-51A2-47F8-9B5C-4284458E47A6") StringTesting : public UnknownImpl, public IStringTesting
+{
+private:
+    template <typename STRING>
+    bool EqualValue(STRING l, STRING r)
+    {
+        STRING tmp = l;
+        int len = 1; // Include 1 for null
+        while (*tmp++)
+            ++len;
+
+        return (0 == ::memcmp(l, r, len * sizeof(l[0])));
+    }
+
+    template <typename STRING>
+    STRING ReverseInplace(size_t len, STRING s)
+    {
+        std::reverse(s, s + len);
+        return s;
+    }
+
+    template<typename STRING>
+    HRESULT Reverse(STRING str, STRING *res)
+    {
+        STRING tmp = str;
+        size_t len = 0;
+        while (*tmp++)
+            ++len;
+
+        size_t strDataLen = (len + 1) * sizeof(str[0]);
+        auto resLocal = (STRING)::CoTaskMemAlloc(strDataLen);
+        if (resLocal == nullptr)
+            return E_OUTOFMEMORY;
+
+        ::memcpy_s(resLocal, strDataLen, str, strDataLen);
+        *res = ReverseInplace(len, resLocal);
+
+        return S_OK;
+    }
+
+    HRESULT ReverseBstr(BSTR str, BSTR *res)
+    {
+        UINT len = ::SysStringLen(str);
+        UINT strDataLen = len * sizeof(str[0]);
+        BSTR resLocal = ::SysAllocStringByteLen(reinterpret_cast<LPCSTR>(str), strDataLen);
+        if (resLocal == nullptr)
+            return E_OUTOFMEMORY;
+
+        *res = ReverseInplace(len, resLocal);
+
+        return S_OK;
+    }
+
+public: // IStringTesting
+    DEF_RAWFUNC(Add_LPStr)(
+        /*[in]*/ LPSTR a,
+        /*[in]*/ LPSTR b,
+        /*[out,retval]*/ LPSTR * pRetVal)
+    {
+        if (a == nullptr || b == nullptr)
+            return E_POINTER;
+
+        size_t aLen = ::strlen(a);
+        size_t bLen = ::strlen(b);
+        auto buf = (LPSTR)::CoTaskMemAlloc((aLen + bLen + 1) * sizeof(*b));
+
+        ::strcpy_s(buf, aLen + 1, a);
+        ::strcpy_s(buf + aLen, bLen + 1, b);
+
+        *pRetVal = buf;
+        return S_OK;
+    }
+    DEF_RAWFUNC(Add_LPWStr)(
+        /*[in]*/ LPWSTR a,
+        /*[in]*/ LPWSTR b,
+        /*[out,retval]*/ LPWSTR * pRetVal)
+    {
+        if (a == nullptr || b == nullptr)
+            return E_POINTER;
+
+        size_t aLen = ::wcslen(a);
+        size_t bLen = ::wcslen(b);
+        auto buf = (LPWSTR)::CoTaskMemAlloc((aLen + bLen + 1) * sizeof(*b));
+
+        ::wcscpy_s(buf, aLen + 1, a);
+        ::wcscpy_s(buf + aLen, bLen + 1, b);
+
+        *pRetVal = buf;
+        return S_OK;
+    }
+    DEF_RAWFUNC(Add_BStr)(
+        /*[in]*/ BSTR a,
+        /*[in]*/ BSTR b,
+        /*[out,retval]*/ BSTR * pRetVal)
+    {
+        if (a == nullptr || b == nullptr)
+            return E_POINTER;
+
+        UINT aLen = ::SysStringLen(a);
+        UINT bLen = ::SysStringLen(b);
+        BSTR buf = ::SysAllocStringByteLen(nullptr, (aLen + bLen) * sizeof(a[0]));
+
+        ::wcscpy_s(buf, aLen + 1, a);
+        ::wcscpy_s(buf + aLen, bLen + 1, b);
+
+        *pRetVal = buf;
+        return S_OK;
```

Comment:
Memory leak: you need to release a, b's memory

Reply:
That is not what I am seeing and in fact when I try to free using `CoTaskMemFree` I get a crash. I have root caused this as some optimizations in the marshler code. It does appear that the marshler attempts to tree `string` as blittable for the callee to minimize issues. This seems right...

Reply:
I verified in the marshlers - strings can handled in a special manner and shouldn't be freed.